### PR TITLE
 C#: Recognize `.proj` files in autobuilder

### PR DESCRIPTION
**Pull request by hvitved**
When determining the target of `msbuild` or `dotnet build`, first look for `.proj` files, then `.sln` files, and finally `.csproj`/`.vcxproj` files. In all three cases, choose the project/solution file closest to the root.

@calumgrant @jf205

### CODE REVIEW

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** and **fixes 1** when merging 1afc529e63a8682a3d8e14efbab4f544db71c158 into 4e2d40aad8d18c3279adfe069927710f5682b9bc - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-fceb44581c294ced18489872e93577dfb7bb4de0)

**new alerts:**

* 1 for Generic catch clause

**fixed alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs [hunk 1]
```diff
@@ -113,46 +119,75 @@ public Autobuilder(IBuildActions actions, AutobuildOptions options)
...
+                        files = allFiles.Select(p => p.ProjectOrSolution).Take(1);
+                        return true;
+                    }
+
+                    var candidates = allFiles.
+                        Where(f => f.DistanceFromRoot == allFiles[0].DistanceFromRoot).
+                        Select(f => f.ProjectOrSolution);
+                    Log(Severity.Info, $"Found multiple '{extension}' files, giving up: {string.Join(", ", candidates)}.");
+                    files = new IProjectOrSolution[0];
+                    return true;
```
Comment (calumgrant) [inline csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs]: This return value is a bit weird - false means carry on and true means stop success or failure.

File: csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs [hunk 2]
```diff
@@ -113,46 +119,75 @@ public Autobuilder(IBuildActions actions, AutobuildOptions options)
...
                 }
 
-                var solutions = GetExtensions(".sln").
-                        Select(s => new Solution(this, s)).
-                        Where(s => s.ProjectCount > 0).
-                        OrderByDescending(s => s.ProjectCount).
-                        ThenBy(s => s.Path.Length).
+                bool FindFiles(string extension, Func<string, ProjectOrSolution> create, out IEnumerable<IProjectOrSolution> files)
+                {
+                    var allFiles = GetExtensions(extension).
```
Comment (calumgrant) [inline csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs]: `allFiles` -> `matchingFiles` ?

File: csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs [hunk 3]
```diff
@@ -113,46 +119,75 @@ public Autobuilder(IBuildActions actions, AutobuildOptions options)
...
                     }
                     return ret;
                 }
 
-                var solutions = GetExtensions(".sln").
-                        Select(s => new Solution(this, s)).
-                        Where(s => s.ProjectCount > 0).
-                        OrderByDescending(s => s.ProjectCount).
-                        ThenBy(s => s.Path.Length).
+                bool FindFiles(string extension, Func<string, ProjectOrSolution> create, out IEnumerable<IProjectOrSolution> files)
```
Comment (calumgrant) [inline csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs]: I don't think this particularly needs to be a local function.

File: csharp/autobuilder/Semmle.Autobuild/BuildActions.cs [hunk 4]
```diff
@@ -96,12 +97,27 @@ public interface IBuildActions
...
 
+        /// <summary>
+        /// Gets the full path for <paramref name="path"/>, Path.GetFullPath().
+        /// </summary>
+        string GetFullPath(string path);
+
+        /// <summary>
+        /// Gets the directory of <paramref name="path"/>, Path.GetDirectoryName().
+        /// </summary>
+        string GetDirectoryName(string path);
```
Comment (calumgrant) [inline csharp/autobuilder/Semmle.Autobuild/BuildActions.cs]: This probably doesn't need to be stubbed out as it doesn't touch the filesystem.

File: csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs (hunk 1 above)
Comment (hvitved) [reply csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs]: The return value behaves like e.g. the return value of `int.TryParse(string s, out int i)`; the output parameter contains valid data iff the return value is `true`.

Comment (hvitved) [reply csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs]: But actually there is no need to have both the Boolean and the enumerable, so I will rewrite.

File: csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs (hunk 3 above)
Comment (hvitved) [reply csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs]: I think local functions are nice (as opposed to private static methods) when they are only used inside one particular callable, as in this case.

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** and **fixes 1** when merging 8648c8e6a00631c1c005a21a1a5d802cb45fac86 into e062851709488e287a6644cba466ec48261dfc1c - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-5b53119c0c964ebac4fd46f4fbdfc8b726a125da)

**new alerts:**

* 1 for Generic catch clause

**fixed alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs [hunk 5]
```diff
@@ -29,26 +29,32 @@ interface IBuildRule
     public class Autobuilder
     {
         /// <summary>
-        /// Full file paths of files found in the project directory.
+        /// Full file paths of files found in the project directory, as well
+        /// their distance from the project root folder. The list is sorted
```
Comment (jf205) [inline csharp/autobuilder/Semmle.Autobuild/Autobuilder.cs]: as well _as_ their distance....?  (same thing for three summaries below)

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** and **fixes 1** when merging 9d0cd1a6eaad42d3d89ec66ebec917c0f477d720 into 201f64ef8ebc8d86506be995cd15a16bbf074d54 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-30008cec4d12ee73ec422167266a2805a427d7ab)

**new alerts:**

* 1 for Generic catch clause

**fixed alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** and **fixes 1** when merging 743fc85ef1ed92027dbc3e2502a183ab358576b7 into 201f64ef8ebc8d86506be995cd15a16bbf074d54 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-24902cf3fe25746de1ae2239af0d0bbe5d44330a)

**new alerts:**

* 1 for Generic catch clause

**fixed alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by hvitved** [comment]
Rebased again.

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** and **fixes 1** when merging 1939773684aab2b619870bbfdac6dec997006ef3 into 4ad59235d8a9308104c42b6dfa711ccba52a71fa - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-b69324c928c3e24d5a993db9a178b03c661af904)

**new alerts:**

* 1 for Generic catch clause

**fixed alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**